### ISSUE TITLE
Remove tini for Docker 1.13 or greater

### ISSUE THREAD

**Issue by harmjanblok** [rook/rook#1667]
Is this a bug report or feature request?
* Feature request

**Feature Request**
We've recently upgraded to docker 1.13.1, and rook Pods now show a warning in the logfiles regarding `tini`:

```
[WARN  tini (7)] Tini is not running as PID 1 and isn't registered as a child subreaper.
Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.
2018-04-25 07:06:57.930568 I | rook: starting Rook v0.6.2 with arguments '/usr/local/bin/rook agent'
2018-04-25 07:06:57.932837 I | rook: flag values: --help=false, --log-level=INFO
2018-04-25 07:06:57.934068 I | rook: starting rook agent
...
```

As described on https://github.com/krallin/tini#using-tini, Tini is included in Docker itself, when you are using Docker 1.13 or greater.

What would be the correct way to handle this and simultaneously provide compatibility with docker 1.12 or older?

**Environment**:
* OS (e.g. from /etc/os-release): CentOS 7
* Kernel (e.g. `uname -a`): 3.10.0-693.17.1.el7.x86_64
* Cloud provider or hardware configuration: vCloudDirector
* Rook version (use `rook version` inside of a Rook Pod): 0.6.2
* Kubernetes version (use `kubectl version`): 1.7.5
* Docker version: 1.13.1

**Comment by BlaineEXE** [comment]
I looked into this a bit myself, and I think with kubernetes 1.7+, there may be no need to use tini because of k8s's new PID namespace sharing enables a zombie reaper for all pods.
https://www.ianlewis.org/en/almighty-pause-container

Please correct me if I am wrong, but I think one-osd-per-pod will require PID namespace sharing, and therefore, we can assume that this feature is enabled. I am not the best to coordinate this, but I think it makes sense that once k8s 1.6 support is dropped and one-osd-per-pod becomes the norm, we can remove tini and simplify Rook's images.

**Comment by tonya11en** [comment]
So it looks like configurable PID namespace sharing is alpha in k8s 1.10 and set for beta in k8s 1.12:
https://github.com/kubernetes/features/issues/495

Since it's going to remain in alpha until 1.12, we can expect people to enable it via a flag to k8s, so it seems we'll need to rip out `tini` for quite a while.

**Comment by jbw976** [comment]
I'm not sure I understand all of this, but I'm trying to piece it together.

So Docker 1.13+ has tini built in and will be used automatically if `--init` flag is passed, e.g., `docker run --init`.  Does Kubernetes pass this flag to `docker run` in the versions Rook supports (1.7+)?

k8s 1.10+ has PID namespace sharing, but it must be turned on with the `PodShareProcessNamespace` feature gate, [as shown here](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/).  Very few deployments will be using this.

In what environments can we safely drop `tini` from our images?  That is not clear to me.

Also, I was not under the impression that 1 OSD per pod (#1656 via #1698) required PID namespace sharing.  @rootfs @travisn can you confirm that please?

**Comment by tonya11en** [comment]
> Does Kubernetes pass this flag to docker run in the versions Rook supports (1.7+)?
No, there's no way to do this, but my understanding is that the PID namespace sharing accomplishes the same thing via the pause container.

> In what environments can we safely drop tini from our images?
No current environment. I misspoke above. We **won't be able to** rip out `tini` for quite a while.
